python/entry_point/generator.py

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python3
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""entry_point generator."""
+
+from __future__ import annotations
+
+import argparse
+import configparser
+import pathlib
+import sys
+import textwrap
+
+_ENTRY_POINTS_TXT = "entry_points.txt"
+_TEMPLATE = """\
+{shebang}
+import sys
+# Drop the first entry in the sys.path, because it will point to our workspace
+# where we are generating the entry_point script and it seems that some package
+# break due to this reason (e.g. pylint). This means that we should not use this
+# script to ever generate entry points for scripts within the main workspace,
+# but that is fine, we can create a separate generator or a boolean flag for
+# that.
```

Comment:
I dug a little deeper and I am confused. The `pylint` fails to start when running via `bazel run` but succeeds when running within the test. I am sure that `rules_python` users would use `pylint` `entry_point` via both mechanisms but I am not sure what exactly is going on and it is getting late.

What I checked:
* The `sys.path` is the same in both instances except for the additional entry to load the `@rules_python//python/runfiles`.
* The entries for the test are using the `sandox` folders whereas `bazel run` is using something else:
  * See `bazel run` syspath here: https://gist.github.com/aignas/396ed948a9a51a59bdb6c4155347c3fe
  * See `bazel test` syspath here: https://gist.github.com/aignas/f909306720d430bcd917444a9791fdc0

And I am not sure why `pylint` starts working if I remove the first `sys.path` entry, possibly because the directory contains the following:
```
.r-xr-xr-x  21k aignas 21 Aug 00:13 entry_point_test
.r-xr-xr-x  34k aignas 21 Aug 00:13 entry_point_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:13 entry_point_test.runfiles
.r-xr-xr-x 1.8M aignas 21 Aug 00:13 entry_point_test.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint
.r-xr-xr-x  14k aignas 21 Aug 00:26 pylint.repo_mapping
drwxrwxrwx    - aignas 21 Aug 00:26 pylint.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint_deps_test
.r-xr-xr-x  20k aignas 21 Aug 00:26 pylint_deps_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 pylint_deps_test.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint_deps_test.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint_test
.r-xr-xr-x  16k aignas 21 Aug 00:26 pylint_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 pylint_test.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint_test.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint_with_deps
.r-xr-xr-x  18k aignas 21 Aug 00:26 pylint_with_deps.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 pylint_with_deps.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint_with_deps.runfiles_manifest
.r-xr-xr-x  619 aignas 21 Aug 00:28 rules_python_entry_point_pylint.py
.r-xr-xr-x  619 aignas 21 Aug 00:28 rules_python_entry_point_pylint_with_deps.py
.r-xr-xr-x  611 aignas 21 Aug 00:28 rules_python_entry_point_yamllint.py
.r-xr-xr-x  21k aignas 21 Aug 00:26 yamllint_test
.r-xr-xr-x 3.4k aignas 21 Aug 00:26 yamllint_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 yamllint_test.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 yamllint_test.runfiles_manifest
```

Hence I am thinking that actually removing the first entry would yield the correct behaviour so that we would not be able to `import` from outside the `runfiles` directory.

Reply:
I have put the workaround as I am somewhat 90% convinced that it is needed but would like to have a better mental model of what the differences are between `bazel run` and `bazel test` in this situation. I have also attempted to create an `alias` and name the underlying `py_binary` something else, but `pylint` was still failing if the first item in the `sys.path` was present.

Reply:
The `bazel run` paths look like what a non-sandboxed test execution would look like which is basically what it is. So that makes sense. Disabling sandbox or forcing local execution would probably make the test look more similar.

Looking at your pastes, I found a couple apparent typos, and 1 potentially meaningful difference:

* The test has 1 extra entry: `pylint_with_deps/runfiles/rules_python~override`, i.e. the bzlmod module root. Is there a directory in there that might be confusing things? From what I recall, all the directories in there have "bzlmod canonical" names (`bla~whatever~blabla`), but don't 100% recall
* Typo-looking things (run vs test):
  * `pip_39_` vs `pip_3_`
  * `python_3_9_x86_64` vs `pyhon_3_9_x86_64`

This all very much sounds like the "sys.path[0] has the script dir" issue that PYTHONSAFEPATH was created to address.

Does this go away when using Bazel at head with Python 3.11? (e.g. USE_BAZEL_VERSION=last_green). That combination should prevent sys.path[0] from being the main py dir.

It looks like sys.flags.isolated and sys.flags.safe_path could be checked to see if the behavior is available/enabled.

Alternatively, try 3.11, set PYTHONSAFEPATH=1 in the env, then re-exec sys.argv and see what sys.path looks like.

Anyways, I think the sys.path[0] removal you have is fine enough.

Reply:
So:
1. I removed the sys.path removal workaround
2. I added pylint with requirements for Python 3.11
3. And then I created a separate target to force usage of Python 3.11
4. Ran `USE_BAZEL_VERSION=last_green bazel run //entry_points:pylint_latest` and it just worked as expected, whereas the `Python 3.9` target was still failing. So I think your analysis is correct here.

Diff used to experiment in the gist: https://gist.github.com/aignas/e5097e93bd7117d95dfe1e557865784a

That lead me to add a line to check about `safe_path`, but I left out `sys.flags.isolated` checks because it was always `0` in my local testing, so we can leave this for a followup if really needed.